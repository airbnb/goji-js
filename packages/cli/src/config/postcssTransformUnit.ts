/* eslint-disable import/no-import-module-exports */
// this plugin is inspired from https://github.com/yingye/postcss-px2units
// and refactored to support PostCSS 8
import type { PluginCreator } from 'postcss';

interface Options {
  divisor: number;
  multiple: number;
  decimalPlaces: number;
  sourceUnit: string;
  targetUnit: string;
  comment: string;
}

const DEFAULT_OPTIONS: Options = {
  divisor: 1,
  multiple: 1,
  decimalPlaces: 2,
  sourceUnit: 'px',
  targetUnit: 'rpx',
  comment: 'no',
};

const postcssTransformUnit: PluginCreator<Partial<Options>> = (options = {}) => {
  const mergedOptions = { ...DEFAULT_OPTIONS, ...options };

  const replacePx = (str: string) => {
    if (!str) {
      return '';
    }

    return str.replace(
      new RegExp(`\\b(\\d+(\\.\\d+)?)${mergedOptions.sourceUnit}\\b`, 'g'),
      (_match, x) => {
        const size = (x * mergedOptions.multiple) / mergedOptions.divisor;
        return size % 1 === 0
          ? size + mergedOptions.targetUnit
          : size.toFixed(mergedOptions.decimalPlaces) + mergedOptions.targetUnit;
      },
    );
  };

  return {
    postcssPlugin: 'postcss-transform-unit',
    Declaration(declaration) {
      const nextNode = declaration.next();
      if (nextNode?.type === 'comment' && nextNode.text === mergedOptions.comment) {
        nextNode.remove();
      } else {
        declaration.value = replacePx(declaration.value);
      }
    },
  };
};

postcssTransformUnit.postcss = true;

module.exports = postcssTransformUnit;
