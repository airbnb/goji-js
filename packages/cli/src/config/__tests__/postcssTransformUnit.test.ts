import postcss from 'postcss';

describe('postcss-transform-unit', () => {
  test('px to rpx', async () => {
    const css = `
      .a {
        width: 10px;
        height: 20rpx;
        font-size: 30px; /* no */
      }
    `;

    const result = await postcss([
      // eslint-disable-next-line global-require
      require('../postcssTransformUnit')({
        divisor: 1,
        multiple: 2,
        sourceUnit: 'px',
        targetUnit: 'rpx',
      }),
    ]).process(css, {
      from: '/path/to/file.css',
    });
    expect(result.css).toMatchSnapshot();
  });

  test('rpx to px', async () => {
    const css = `
      .a {
        width: 10px;
        height: 20rpx;
        font-size: 30rpx; /* no */
      }
    `;

    const result = await postcss([
      // eslint-disable-next-line global-require
      require('../postcssTransformUnit')({
        divisor: 2,
        multiple: 1,
        sourceUnit: 'rpx',
        targetUnit: 'px',
      }),
    ]).process(css, {
      from: '/path/to/file.css',
    });
    expect(result.css).toMatchSnapshot();
  });
});
