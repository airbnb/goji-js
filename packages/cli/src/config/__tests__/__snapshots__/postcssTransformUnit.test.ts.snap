// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postcss-transform-unit px to rpx 1`] = `
"
      .a {
        width: 20rpx;
        height: 20rpx;
        font-size: 30px;
      }
    "
`;

exports[`postcss-transform-unit rpx to px 1`] = `
"
      .a {
        width: 10px;
        height: 10px;
        font-size: 30rpx;
      }
    "
`;
